### ISSUE TITLE
Maximum recursion depth

### ISSUE THREAD

**Issue by gabbersepp** [AngleSharp/AngleSharp#763]
### Description

Add possibility to restrict recursion depth to a maximum number.

### Background

We use AngleSharp to parse E-Mails. Those can come from good people and from bad (Spam). In the latter case, the E-Mails often contain invalid tags and invalid HTML.
Currently we run into issues with following stuff:

```
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
<div></div><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV=""><D IV="">
```
And this is only a small piece. The original email is about ten times longer.
This one leads to a StackOverflow in `CollectionExtensions.GetAllElements<T>(this INode parent, Predicate<T> predicate)` while accessing child nodes.

Of course I know that the maximum depth depends on several factors, e.g. from where the code is called and how much is on the stack.

Any other suggestions to solve this issue? 

//Edit: maybe we are able to use a iterative solution instead of a recursive one. Are there any reasons for the current choice?

**Comment by FlorianRappl** [comment]
You can always transform a recursive algorithm into an iterative one, however, the readability usually suffers.

In the given case (not sure if I remember) there is no simple mapping, so you would need to keep track of the elements via a stack (what a call stack gives you out of the box). Naturally, you can go there, but this stack is also limited by memory (not as much as the call stack, but of course there are physical limits) and not as efficient / optimized. Long story short: I think you will exchange one problem with another (in this simple transformation).

TL;DR: If there is a good solution algorithm any contribution would be welcome. There are many spots - this is just one of them.

**Comment by campersau** [comment]
For anybody who is interested in fixing these, I have done this before in https://github.com/AngleSharp/AngleSharp/pull/524, https://github.com/AngleSharp/AngleSharp/pull/526 and https://github.com/AngleSharp/AngleSharp/pull/544

Here is the code @FlorianRappl was taking about: https://github.com/campersau/AngleSharp/blob/7ef760cc6c4eb4f19e306038cbde4c0b64636a43/src/AngleSharp/Dom/NodeExtensions.cs#L113-L128

**Comment by gabbersepp** [comment]
Currently I do not know what @FlorianRappl means by "simple mapping" in this case. 
I will have to investigate the source code and the commits from @campersau in the next few days to fully understand the problem here. 

But if it holds that an iterative solution is creating another problem, maybe we can inject a predicate through configuration?

**Comment by campersau** [comment]
The best solution would be an algorithm like this https://github.com/AngleSharp/AngleSharp/pull/524#issuecomment-276221408 which doesn't use a stack at all but it is very slow for `Anglesharp` because of the current [`NextSibling` implementation](https://github.com/AngleSharp/AngleSharp/blob/ba33f7ebbe5ebfb2bdb533d51fcd5a9c688c9731/src/AngleSharp/Dom/Internal/Node.cs#L182-L201).

**Comment by gabbersepp** [comment]
@campersau I am not sure If I am missing the point. But why is the NextSibling slow? If I have calculated correctly, this should be an O(n/2) algorithm which is not very bad.

**Comment by campersau** [comment]
I can remember that I had changed one of the performance projects to do a small perf tests on both of theses implementations and the implementation with `NextSibling` was a lot slower. Sadly I don't have the exact numbers anymore.
The problem here is that it is a loop inside of another loop which makes it so slow. E.g. if you have a table with a lot of rows than for each row you would need to find the next sibling which in turn loops through the rows of the table each time.

**Comment by gabbersepp** [comment]
@campersau @FlorianRappl what do you think about this static update of NextSiblings?
https://github.com/gabbersepp/AngleSharp/commit/5dce81a24ce71afb3eabd072f5409806847e9aba
Tests are green but I have no deep knowledge about this project so maybe you easily can tell me whether or not I should make a pull request from this commit.

**Comment by FlorianRappl** [comment]
@gabbersepp interesting idea, but I think this will invite some bugs and oddities + it is now inconsistent with `PreviousSibling` and the respective `Element` counterparts. However, worst part is additional memory consumption, which is now +1 field, but with all the described support would be +4 (or even more) fields per node.

Hence I am not sure if we should pursue this path.

**Comment by gabbersepp** [comment]
@FlorianRappl It is a pointer to a Node which should take 4 Bytes on a 64Bit machine.
If we use a index based solution we can reduce it to 2 bytes per field. One byte would allow too little siblings. 
Previous-/NextSiblings should be able to be combined into one index field if every Node hold it's current index in the list of parent's siblings.
You're right, there may inconsistencys right now. I have not search through the code for similar usages. :-)

**Comment by FlorianRappl** [comment]
> It is a pointer to a Node which should take 4 Bytes on a 64Bit machine.

Do you mean 8 bytes (4 bytes on a 32-bit machine)?

If we have larger trees (and I guess this is exactly your target) then we talk about 1k - 10k nodes minimum, which means 4 (Prev, Next for Node and Element) * 8 * 1000 = 32k memory added for a document with 1k nodes (plus potential leaks as the GC suffers from the direct references, which means one needs to be super careful not only for obvious bugs, but also for non-obvious ones).

Don't get me wrong here - all suggestions are viable. I just want to point out that any change on `Node` level directly influences memory (as `Node` is the bread and butter in AngleSharp). So even a simple change like adding 1 field can have a huge impact.

Experience in AngleSharp shows that freeing nodes with recursive dependencies is also handled worse by the GC than it should be.

Alternative could be to have the child nodes implemented as a linked list. It is a lower level construct and therefore consistent by design. However, while the linked list would improve the speed on the traversal there, it will have other drawbacks.

**Comment by gabbersepp** [comment]
Ah yes you're right. 8 bytes per reference.
Ok. I understand the problem.
I have thought a bit about this problem without success.
One optimization that can be done is to replace the "name" field with an index reference to a global list of tag names. 16 bit should be enhough and thus save 3/4 of the current field's size.
Searching for the correct location within all tag names should not take more time than now I think as .NET will do something similar when doing string interning.
This can compensate one of the mentioned fields :-D

**Comment by gabbersepp** [comment]
Wanted to note that until a good solution is found, I am using this code to check the depth:

```csharp
        public static IHtmlDocument SaveParse(this HtmlParser parser, string content)
        {
            // ReSharper disable once AngleSharpStackOverflowCheck
            var parsed = parser.Parse(content);

            foreach (var e in parsed.All)
            {
                INode node = e;
                int counter = 0;

                while ((node = node.Parent) != null)
                {
                    counter++;

                    if (counter > 500)
                    {
                        throw new Exception("tree max depth reached. Skip to avoid SO");
                    }
                }
            }

            return parsed;
        }
```

Runtime should be round about O(nlog(n)) for balanced html documents and n^2 for those "newsletters of death". I will make a performance test with a bunch of existing emails in our DB and update this comment with the results.

**Comment by gabbersepp** [comment]
Performance statistics:

email count | error | success | ms | ticks | length of all emails | size / email | time / email | error / day 
------------- | ----- | --------- | --- | ----- | --------------------- | ------------ | ------------- | ------------ 
800142 | 158 | 799984 | 549297 | 1454777529 | 6160514531 | 7699 | 0.68ms | 0.114

**Comment by FlorianRappl** [comment]
I'll try to come up with a solution in unlucky number 13.

**Comment by FlorianRappl** [comment]
I am currently looking into the problem and I have problems seeing where the issue is ...

The `GetAllNodes` extension method (I think this is the one you refer to) seems not to be recursive. Instead, it uses `NodeEnumerator` which has a special `MoveNext` implementation. `MoveNext` is called when iterating over `foreach` loops.

Can you provide such an email of death with a small code that you user to trigger the SO exception? Thanks!

**Comment by gabbersepp** [comment]
@FlorianRappl Hey I will send you an example next week.

**Comment by gabbersepp** [comment]
@FlorianRappl please see this gist:
https://gist.github.com/gabbersepp/033b5a1fa93f055a8b1d816db5cdcaf3

I have removed all sensible information.

**Comment by FlorianRappl** [comment]
Great, thanks!

**Comment by FlorianRappl** [comment]
Hm I got a green test for it ...

```c#
[Test]
public void IteratingThroughDomShouldNeverFail_Issue763()
{
  var content = Assets.GetManifestResourceString("Html.BrokenMail.txt");
  var p = new HtmlParser().ParseDocument(content);
  Assert.IsNotNull(p);
  var img = p.Images.Select(m => m).ToList();
  Assert.AreEqual(1, img.Count);
}
```

Anything I'm missing @gabbersepp ?

**Comment by gabbersepp** [comment]
@FlorianRappl 
Please try to use the max stack size as shown in the gist. It is a bit weird. If I use this email in out ASP.NET WebAPI projekt, a stackoverflow is thrown. If I copy the same code into a console app, nothing happens. To force the SO, I have to reduce the stack size.

**Comment by FlorianRappl** [comment]
Apparently, the stack trace has not the "big" impact (even though a larger stack potentially solves it, too) - its the `var html = p.ToHtml();` which makes it crash.

**Comment by FlorianRappl** [comment]
I flattened the `ToHtml` implementation. Landed in `devel`.
